Purchases/SwiftSources/Purchasing/ProductsManager.swift

```diff
@@ -0,0 +1,99 @@
+//
+//  ProductsManager.swift
+//  Purchases
+//
+//  Created by Andrés Boedo on 7/14/20.
+//  Copyright © 2020 Purchases. All rights reserved.
+//
+
+import Foundation
+import StoreKit
+
+internal class ProductsManager: NSObject {
+    private let productsRequestFactory: ProductsRequestFactory
+
+    private var cachedProductsByIdentifier: [String: SKProduct] = [:]
+    private let queue = DispatchQueue(label: "ProductsManager")
+    private var productsByRequests: [SKRequest: Set<String>] = [:]
+    private var completionHandlers: [Set<String>: [(Set<SKProduct>) -> Void]] = [:]
+
+    init(productsRequestFactory: ProductsRequestFactory = ProductsRequestFactory()) {
+        self.productsRequestFactory = productsRequestFactory
+    }
+
+    func products(withIdentifiers identifiers: Set<String>, completion: @escaping (Set<SKProduct>) -> Void) {
+        queue.async { [self] in
+            let productsAlreadyCached = self.cachedProductsByIdentifier.filter { key, _ in identifiers.contains(key) }
+            if productsAlreadyCached.count == identifiers.count {
+                let productsAlreadyCachedSet = Set(productsAlreadyCached.values)
+                NSLog("skipping products request because products were already cached. products: \(identifiers)")
```

Comment:
We are using `NSLog` in this file. Should we copy the logs utils we have to be able to use them in this swift files?

Reply:
once we get the new RCLog stuff Tina's working on merged, we can come back and refactor this to use it